thewalrus/tests/test_fock_gradients.py

```diff
@@ -148,25 +124,122 @@ def test_Kgate():
+
+def test_Dgate_values(tol):
+    """Tests the correct construction of the single mode displacement operation"""
+    cutoff = 5
+    alpha = 0.3 + 0.5 * 1j
+    # This data is obtained by using qutip
+    # np.array(displace(40,alpha).data.todense())[0:5,0:5]
+    expected = np.array(
+        [
+            [0.84366482 + 0.00000000e00j, -0.25309944 + 4.21832408e-01j, -0.09544978 - 1.78968334e-01j, 0.06819609 + 3.44424719e-03j, -0.01109048 + 1.65323865e-02j,],
+            [0.25309944 + 4.21832408e-01j, 0.55681878 + 0.00000000e00j, -0.29708743 + 4.95145724e-01j, -0.14658716 - 2.74850926e-01j, 0.12479885 + 6.30297236e-03j,],
+            [-0.09544978 + 1.78968334e-01j, 0.29708743 + 4.95145724e-01j, 0.31873657 + 0.00000000e00j, -0.29777767 + 4.96296112e-01j, -0.18306015 - 3.43237787e-01j,],
+            [-0.06819609 + 3.44424719e-03j, -0.14658716 + 2.74850926e-01j, 0.29777767 + 4.96296112e-01j, 0.12389162 + 1.10385981e-17j, -0.27646677 + 4.60777945e-01j,],
+            [-0.01109048 - 1.65323865e-02j, -0.12479885 + 6.30297236e-03j, -0.18306015 + 3.43237787e-01j, 0.27646677 + 4.60777945e-01j, -0.03277289 + 1.88440656e-17j,],
+        ]
+    )
+    T = Dgate(np.abs(alpha), np.angle(alpha), cutoff)[0]
+    assert np.allclose(T, expected, atol=tol, rtol=0)
+
+
+def test_Sgate_values_complex(tol):
+    """Tests the correct construction of the single mode squeezing operation"""
+    r = 0.5
+    theta = 0.7
+    cutoff = 5
+    # This data is obtained by using qutip
+    # np.array(squeeze(40,0.5*np.exp(1j*0.7)).data.todense())[0:5,0:5]
+    expected = np.array(
+        [
+            [0.94171062 + 0.0j, 0.0 + 0.0j, 0.23535661 - 0.19823814j, 0.0 + 0.0j, 0.02093159 - 0.12135894j,],
+            [0.0 + 0.0j, 0.83512676 + 0.0j, 0.0 + 0.0j, 0.36151137 - 0.30449682j, 0.0 + 0.0j],
+            [-0.23535661 - 0.19823814j, 0.0 + 0.0j, 0.64005396 + 0.0j, 0.0 + 0.0j, 0.42261153 - 0.35596078j,],
+            [0.0 + 0.0j, -0.36151137 - 0.30449682j, 0.0 + 0.0j, 0.38926873 + 0.0j, 0.0 + 0.0j],
+            [0.02093159 + 0.12135894j, 0.0 + 0.0j, -0.42261153 - 0.35596078j, 0.0 + 0.0j, 0.12407853 + 0.0j,],
+        ]
+    )
+    T = Sgate(r, theta, cutoff)[0]
+    assert np.allclose(T, expected, atol=tol, rtol=0)
+
+
+def test_BS_selection_rules(tol):
+    r"""Test the selection rules of a beamsplitter.
+    If one writes the beamsplitter gate of :math:`U` and its matrix elements as
+    :math:`\langle m, n |U|k,l \rangle` then these elements
+    are nonzero if and only if :math:`m+n = k+l`. This test checks
+    that this selection rule holds.
+    """
+    cutoff = 4
+    T = BSgate(np.random.rand(), np.random.rand(), cutoff)[0]
+    for m in range(cutoff):
+        for n in range(cutoff):
+            for k in range(cutoff):
+                for l in range(cutoff):
+                    if m + n != k + l:  # Check that there are the same total number of photons in the bra and the ket
+                        assert np.allclose(T[m, n, k, l], 0.0, atol=tol, rtol=0)
```

Comment:
I'm wondering if there is a way to write this test in a vectorised manner?

```python
m = np.arange(5).reshape(-1, 1, 1, 1)
n = np.arange(5).reshape(1, -1, 1, 1)
k = np.arange(5).reshape(1, 1, -1, 1)
l = np.arange(5).reshape(1, 1, 1, -1)

# create a copy of T, but replace all elements where
# m+n != k+l with 0.
S = np.where(m + n != k + l, 0, T)

# check that S and T remain equal
assert np.allclose(S, T, atol=tol, rtol=0)
```

The advantage here is that if the test fails, since the assert is not inside a loop, it is easy to see from the traceback which element of the tensor is incorrect.

Reply:
This looks elegant. I will implement it. Thanks @josh146 !

Reply:
Done!

Reply:
:+1: maybe an even better approach would be to use the m, n, l, k arrays to extract the elements of T that satisfy the selection rule, and assert all elements are equal. But couldn't think how to do that while on the plane!